misc/bootstrap.sh

```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+set -eEu
+set -o pipefail
+. test/util.sh
+
+################################################################################
+# Configure your local dev environment for this git repo.
+# Invoke as "misc/bootstrap".
+################################################################################
+
+# Update git submodules.
+smitty git submodule sync
+smitty git submodule update --init --recursive
+
+# Clean/reset git submodules.
+warn "About to clean git submodules."
+echo "Press ^C within 3s to abort."
+sleep 3s
+git clean -ffd
+
+# Add an "upstream" git remote.
+UPSTREAM_URI="https://github.com/purpleidea/mgmt.git"
+if git remote show upstream 2> /dev/null | grep "${UPSTREAM_URI}" &> /dev/null; then
+	info upstream git remote is OK
+else
+	git remote rm upstream &> /dev/null || :
+	warn "Setting upstream remote."
+	smitty git remote add upstream "${UPSTREAM_URI}"
+fi
+
+# Allow to easily checkout Pull Requests (PRs) locally.
+if [[ "$(git config --list)" =~ remote.upstream.fetch=\+refs/pull/\*/head:refs/remotes/upstream/pr/\* ]]; then
+	info config to fetch upstream PR is OK
+else
+	warn "Configuring git to fetch pull requests. See \"misc/checkout-pr.sh\" for info."
+	smitty git config --add remote.upstream.fetch '+refs/pull/*/head:refs/remotes/upstream/pr/*'
+fi
+
+# Install a specific version of gometalinter.
+go get -u gopkg.in/alecthomas/gometalinter.v1
+dir="$(dirname $(which gometalinter.v1))"
+mv "${dir}/gometalinter.v1" "${dir}/gometalinter"
+
+# Install or update go dependencies.
+smitty go get golang.org/x/tools/cmd/goimports
```

Comment:
what's a smitty btw?

Reply:
it's a function that shows a command and its arguments, then runs the command.

i like to use it for wrapper scripts to passively educate users about what's happening behind the scenes. many users ignore it, but others see the commands and get curious.

example:

```
$ misc/bootstrap.sh 
[INFO] Running misc/bootstrap.sh
[INFO] git submodule sync
Synchronizing submodule url for 'vendor/github.com/coreos/etcd'
Synchronizing submodule url for 'vendor/github.com/grpc-ecosystem/go-grpc-prometheus'
Synchronizing submodule url for 'vendor/github.com/grpc-ecosystem/grpc-gateway'
Synchronizing submodule url for 'vendor/github.com/ugorji/go'
Synchronizing submodule url for 'vendor/google.golang.org/grpc'
Synchronizing submodule url for 'vendor/gopkg.in/fsnotify.v1'
Synchronizing submodule url for 'vendor/honnef.co/go/augeas'
[INFO] git submodule update --init --recursive
[WARN] About to clean git submodules.
Press ^C within 3s to abort.
[INFO] git clean -ffd
[INFO] upstream git remote is OK
[INFO] config to fetch upstream PR is OK
[INFO] go get golang.org/x/tools/cmd/goimports
[INFO] go get github.com/alecthomas/gometalinter
-snip-
```

It's an homage to AIX's [SMIT](https://www.ibm.com/developerworks/aix/library/au-smit/index.html), which has a gui for a new admin to click around in, then it shows the actual command and args. AIX has a tui version called [smitty](https://www.ibm.com/support/knowledgecenter/en/ssw_aix_71/com.ibm.aix.cmds5/smitty.htm).

I've always felt the AIX technique to allow a new admin to use a gui, but then show the actual CLI version, to be a useful teaching technique.

:-)